Part 2 of system for builtin objects

hubertp opened the pull request:
### Pull Request Description

This is the 2nd part of DSL improvements that allow us to generate a lot of
builtins-related boilerplate code.
- [x] generate multiple method nodes for methods/constructors with varargs
- [x] expanded processing to allow for @Builtin to be added to classes and
       and generate @BuiltinType classes
- [x] generate code that wraps exceptions to panic via `wrapException`
        annotation element (see @Builtin.WrapException`

Also rewrote @Builtin annotations to be more structured and introduced some nesting, such as
@Builtin.Method or @Builtin.WrapException.

This is part of incremental work and a follow up on https://github.com/enso-org/enso/pull/3444.

### Important Notes

Notice the number of boilerplate classes removed to see the impact.
For now only applied to `Array` but should be applicable to other types.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

hubertp committed:
Generate BuiltinType from @Builtin

Expanded processing to allow for @Builtin to be added to classes.
More boilerplate gone.

hubertp committed:
Generate method nodes for varargs

When a method or constructor has vararg parameter, we may want to
generate multiple BuiltinMethod representation, up to some number of
parameters.
The approach has been extended in a generic way to take it into
consideration when `expandVargars` parameter is bigger than zero.

This allowed us to remove 4 more hardcoded classes.

hubertp committed:
Generate code that wraps exceptions to panic

Added `wrapException` annotation element that allows us to catch and
wrap runtime exceptions into panic.
`Array.at` and `Array.set_at` nodes are now generated from simple methods.

Additionally, we also have to propagate user-defined annotations on
parameters.

hubertp committed:
Merge branch 'develop' into wip/hubert/builtin-objects-part-2-181499077

hubertp committed:
Formatting

hubertp committed:
Add CHANGELOG

hubertp committed:
Fix typo

JaroslavTulach commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinsProcessor.java:
```diff
@@ -52,36 +62,143 @@ public final boolean process(Set<? extends TypeElement> annotations, RoundEnviro
+    generateBuiltinType(gen, builtinType, stdLibName);
+  }
+
+  private void generateBuiltinType(
```
It might be interesting to see the generated output for a particular usage. It is hard to envision that without actually trying to compile the PR.

hubertp replied:
You mean in the comments? Because we don't show anywhere that for any processor. The generated classes are essentially your source of truth

hubertp replied:
I added a lot of documentation as part of the rewrite I did  to the DSL. Hopefully that improves the situation.

hubertp committed:
Rewrite DSL, allow for repeated exception wrapping

The existing Builtin DSL was becoming a bit too crowded with multiple
elements and default values. That wouldn't scale so had to refactor it.

The functionality stays almost the same except that now we really can
have repeatable @WrapException annotation indicating multiple
catch-clauses for possible exceptions that need to be wrapped.

Added a ton of documentation as the implenetation of the processor is
getting a bit more complicated.

hubertp committed:
Merge branch 'develop' into wip/hubert/builtin-objects-part-2-181499077

hubertp commented:
@4e6 ping

hubertp committed:
Apply DSL to Ref methods

hubertp committed:
Make formatter happy

hubertp committed:
Fix a problem in separate compilation

Discovered a problem with separate compilation when generating exception
wrappers. The latter requires information about all builtin types
annotations but that is not present in separate compilation.
Instead we do a similar trick as in other processors - we read metadata
and diff the results.

The change is bigger than a single method because I didn't want to write
a yet another parser of metadata entries. Instead, every annotation
processor has to provide that implementation and it can be re-used.

hubertp committed:
Fix formatting

hubertp committed:
Merge branch 'develop' into wip/hubert/builtin-objects-part-2-181499077

4e6 commented on build.sbt:
```diff
@@ -157,10 +157,11 @@ Global / onChangedBuildSource := ReloadOnSourceChanges
+  "UTF-8",             // Specify character encoding used by Java source files
+  "-deprecation",      // Shows a description of each use or override of a deprecated member or class
+  "-g",                // Include debugging information
+  "-Xlint:unchecked",  // Enable additional warnings
```
I really like it

kustosz commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -30,6 +32,7 @@ public class Array implements TruffleObject {
    *
    * @param items the element values
    */
+  @Builtin.Method(expandVarargs = 4, description = "Creates an array with given elements.")
```
Ooof. I think this is pointless. The `new_{n}`  family were just helpers for the java-side tests for when they had no access to stdlib. We should have got rid of them a loooong time ago. And it seems like the whole `expandVarargs` functionality is basically useless otherwise? At least I can't think of another case.

kustosz commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -104,6 +105,19 @@ long getArraySize() {
   }
 
+  @Builtin.Method(name = "at", description = "Gets an array element at the given index.")
+  @Builtin.WrapException(from = IndexOutOfBoundsException.class, to = InvalidArrayIndexError.class)
```
holy cow these are such a cool idea! :O

kustosz reviewed (commented):
The varargs part of this seems unnecessary, but I love the WrapException stuff!
